vocabularies/Common.xml

```diff
@@ -1454,6 +1454,132 @@ If the request succeeds the response will contain a JSON body of `Content-Type:
+        <Annotation Term="Core.Description" String="Template for functions that compute AI-based recommendations for the bound entity and its related entities" />
+        <Annotation Term="Core.LongDescription">
+          <String>If the function is invoked repeatedly in a stateful session, it SHOULD NOT
+recompute recommendations whose predictors have not changed since the previous invocation.
+
+The function computes the recommendations asynchronously and returns a URL where its result can be retrieved.
+
+The template function itself cannot be invoked.</String>
+        </Annotation>
+        <Parameter Name="Entity" Type="Edm.EntityType" Nullable="false" />
+        <Parameter Name="Targets" Type="Collection(Edm.String)" Nullable="false">
+          <Annotation Term="Core.IsURL" />
+          <Annotation Term="Core.Description" String="URLs (relative to the request URL) that address entities, properties or action/function parameters for which recommendations shall be retrieved" />
+          <Annotation Term="Core.LongDescription">
+            <String>The request `SalesOrder('A')/ns.RecommendationsFunction?Targets=["","Items(10)"]`
+            retrieves recommendations for the entities `SalesOrder('A')` and `SalesOrder('A')/Items(10)`.</String>
+          </Annotation>
+        </Parameter>
+        <ReturnType Type="Edm.String" Nullable="false">
+          <Annotation Term="Core.IsURL" />
+          <Annotation Term="Core.Description" String="URL (relative to the service URL) for retrieving the recommendations (which are computed asynchronously)" />
+          <Annotation Term="Core.Example">
+            <Record Type="Core.PrimitiveExampleValue">
+              <PropertyValue Property="Description" String="Relative URL for invoking a function import of the [retrieval function](#Template_RecommendationsRetrievalFunction)" />
+              <PropertyValue Property="Value" String="RetrieveRecommendations?ID=af43c2" />
+            </Record>
+          </Annotation>
+        </ReturnType>
+      </Function>
+      <Function Name="Template_RecommendationsRetrievalFunction">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="Template for functions that retrieve AI-based recommendations for the bound entity and its related entities" />
+        <Annotation Term="Core.LongDescription">
+          <String>The template function itself cannot be invoked.</String>
+        </Annotation>
+        <Parameter Name="ID" Type="Edm.String" Nullable="false">
+          <Annotation Term="Core.Description" String="ID that occurs in the URL returned by a prior invocation of the [recommendations function](#Template_RecommendationsFunction)" />
+        </Parameter>
+        <ReturnType Type="Collection(Common.PropertyRecommendationType)" Nullable="false">
+          <Annotation Term="Core.Description" String="A collection of recommendations per targeted property" />
+        </ReturnType>
```

Comment:
Instead of a table with generic return type

target|value|text|alternatives
---|---|---|---
`Items(10)/Product`|M10|Milk|...
`Items(20)/Product`|M11|Cheese|...

wouldn't the following return value be easier to consume for the client?

```
{"Items": [
  {"Item": "010", "Product": {
    "value": "M10", "text": "Milk", "alternatives": ...
  }},
  {"Item": "020", "Product": {
    "value": "M11", "text": "Cheese", "alternatives": ...
  }}
]}
```

The return type of the recommendations retrieval function would then be entity-specific, but could be generated by the SADL designtime. And the `"value"` could have the correct type (for example, `Edm.Date`), not just `Edm.String`.

Reply:
To be discussed in Albatross meeting.

Reply:
No need for this change.